src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,754 @@
+	case SOF_IPC_FRAME_S32_LE:
+		res = MAXX_BUFFER_FORMAT_Q1_31;
+		break;
+	case SOF_IPC_FRAME_FLOAT:
+		res = MAXX_BUFFER_FORMAT_FLOAT;
+		break;
+	default:
+		res = -EINVAL;
+		break;
+	}
+	return res;
+}
+
+/* convert sof frame format to MaxxBuffer_Layout_t */
+static MaxxBuffer_Layout_t layout_convert_sof_to_me(uint32_t layout)
+{
+	MaxxBuffer_Layout_t res;
+
+	switch (layout) {
+	case SOF_IPC_BUFFER_INTERLEAVED:
+		res = MAXX_BUFFER_LAYOUT_INTERLEAVED;
+		break;
+	case SOF_IPC_BUFFER_NONINTERLEAVED:
+		res = MAXX_BUFFER_LAYOUT_DEINTERLEAVED;
+		break;
+	default:
+		res = -EINVAL;
+		break;
+	}
+	return res;
+}
+
+/* check if sample format supported by codec */
+static bool format_is_supported(enum sof_ipc_frame format)
+{
+	bool supported;
+
+	switch (format) {
+	case SOF_IPC_FRAME_S16_LE:
+	case SOF_IPC_FRAME_S24_4LE:
+	case SOF_IPC_FRAME_S32_LE:
+		supported = true;
+		break;
+	case SOF_IPC_FRAME_FLOAT:
+	default:
+		supported = false;
+		break;
+	}
+	return supported;
+}
+
+/* check if buffer layout supported by codec */
+static bool layout_is_supported(uint32_t layout)
+{
+	bool supported;
+
+	switch (layout) {
+	case SOF_IPC_BUFFER_INTERLEAVED:
+		supported = true;
+		break;
+	case SOF_IPC_BUFFER_NONINTERLEAVED:
+	default:
+		supported = false;
+		break;
+	}
+	return supported;
+}
+
+/* check if sample rate supported by codec */
+static bool rate_is_supported(uint32_t rate)
+{
+	bool supported;
+
+	switch (rate) {
+	case 44100:
+	case 48000:
+		supported = true;
+		break;
+	default:
+		supported = false;
+		break;
+	}
+	return supported;
+}
+
+/* allocate memory for MaxxEffect object */
+static int waves_effect_allocate(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = codec->private;
+	MaxxStatus_t status;
+
+	comp_dbg(dev, "waves_effect_allocate() start");
+
+	status = MaxxEffect_GetEffectSize(&waves_codec->effect_size);
+	if (status) {
+		comp_err(dev, "waves_effect_allocate() MaxxEffect_GetEffectSize() error %d",
+			 status);
+		return -EINVAL;
+	}
+
+	waves_codec->effect = (MaxxEffect_t *)codec_allocate_memory(dev,
+		waves_codec->effect_size, 16);
+
+	if (!waves_codec->effect) {
+		comp_err(dev, "waves_effect_allocate() allocate %d bytes for effect",
+			 waves_codec->effect_size);
+		return -ENOMEM;
+	}
+
+	comp_info(dev, "waves_codec_init() allocated %d bytes for effect",
+		  waves_codec->effect_size);
+
+	comp_dbg(dev, "waves_codec_init() done");
+	return 0;
+}
+
+/* checks if sink/source parameters fit MaxxEffect */
+static int waves_effect_check(struct comp_dev *dev)
+{
+	struct comp_data *component = comp_get_drvdata(dev);
+	const struct audio_stream *src_fmt = &component->ca_source->stream;
+	const struct audio_stream *snk_fmt = &component->ca_sink->stream;
+
+	comp_dbg(dev, "waves_effect_check() start");
+
+	/* resampling not supported */
+	if (src_fmt->rate != snk_fmt->rate) {
+		comp_err(dev, "waves_effect_check() sorce %d sink %d rate mismatch",
```

Comment:
Minor comment: you technically don't need to put the function name since SOF logger can identify line numbers for each log message and the function can be easily found once you have the file and line number. Just a FYI (you can keep the function name in the message if you want to)

Reply:
@paulstelian97 putting the function name looks to be the norm in SOF so I think it is fine.

Reply:
typo fixed, thanks
As for function names in logging strings, they come in pretty handy when using grep on sof-logger output